addons/web_editor/static/src/components/media_dialog/media_dialog.js

```diff
@@ -171,33 +173,27 @@ export class MediaDialog extends Component {
                     element.setAttribute('style', style);
                 }
                 if (this.state.activeTab === TABS.IMAGES.id) {
-                    if (this.props.media.dataset.shape) {
-                        element.dataset.shape = this.props.media.dataset.shape;
-                    }
-                    if (this.props.media.dataset.shapeColors) {
-                        element.dataset.shapeColors = this.props.media.dataset.shapeColors;
-                    }
-                    if (this.props.media.dataset.shapeFlip) {
-                        element.dataset.shapeFlip = this.props.media.dataset.shapeFlip;
-                    }
-                    if (this.props.media.dataset.shapeRotate) {
-                        element.dataset.shapeRotate = this.props.media.dataset.shapeRotate;
-                    }
-                    if (this.props.media.dataset.hoverEffect) {
-                        element.dataset.hoverEffect = this.props.media.dataset.hoverEffect;
-                    }
-                    if (this.props.media.dataset.hoverEffectColor) {
-                        element.dataset.hoverEffectColor = this.props.media.dataset.hoverEffectColor;
-                    }
-                    if (this.props.media.dataset.hoverEffectStrokeWidth) {
-                        element.dataset.hoverEffectStrokeWidth = this.props.media.dataset.hoverEffectStrokeWidth;
-                    }
-                    if (this.props.media.dataset.hoverEffectIntensity) {
-                        element.dataset.hoverEffectIntensity = this.props.media.dataset.hoverEffectIntensity;
-                    }
-                    if (this.props.media.dataset.shapeAnimationSpeed) {
-                        element.dataset.shapeAnimationSpeed = this.props.media.dataset.shapeAnimationSpeed;
+                    const oldImageData = registry.category("image.data").get(this.props.media.getAttribute("src"), {});
+                    const newImageData = {};
+                    const optionsToSave = ["shape", "shapeColors", "shapeFlip", "shapeRotate",
+                        "hoverEffect", "hoverEffectColor", "hoverEffectStrokeWidth",
+                        "hoverEffectIntensity", "shapeAnimationSpeed"];
+                    for (const option of optionsToSave) {
+                        if (oldImageData[option]) {
+                            newImageData[option] = oldImageData[option];
+                            if (weUtils.isDatasetImageOption(option)) {
+                                // Options to put in the dataset
+                                element.dataset[option] = oldImageData[option];
+                            }
+                        }
                     }
+                    // Update the "pre.image.data" registry with options that
+                    // have to be kept from the old image.
+                    registry.category("pre.image.data").add(element.getAttribute("src"), newImageData, {force: true});
```

Comment:
Not sure of the best way to do here: I used another registry (`pre.image.options`) as I did not want to fill the `image.options` registry with the options. Indeed, the key-value pair would be wrong as the key would be the original source on which options should be applied in the future. However, the problem with the `pre.image.options` system is that the system can not rely on the correctness of the `image.options` entry from this moment up to when the pre-options are applied on the image and the src of the image is changing (thanks to `_onImageChanged()` of the `ImageTools` option). This is a problem as it is typically in this gap that the `_computeVisibility()` method of the `GridImage` option is executed. This method can so not rely on the `image.options` registry to process a potential shape on the image.

Reply:
Seems louche indeed. Maybe the previous code itself was flawed entirely and we should review the way it works. E.g. why can't the final image be computed here? (I don't know this code, just a thing to investigate). Speak with the reviewer I'll delegate this to for when I am away :+1: